staging/src/k8s.io/client-go/applyconfigurations/meta/v1/unstructured.go

```diff
@@ -0,0 +1,130 @@
+package v1
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+// UnstructuredExtractor enables extracting the applied configuration state from object for fieldManager into an
+// unstructured object type.
+type UnstructuredExtractor interface {
+	ExtractUnstructured(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error)
+	ExtractUnstructuredStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error)
+}
+
+// gvkParserCache caches the GVKParser in order to prevent from having to repeatedly
+// parse the models from the open API schema when the schema itself changes infrequently.
+type gvkParserCache struct {
+	// discoveryClient is the client for retrieving the openAPI document and checking
+	// whether the document has changed recently
+	discoveryClient discovery.DiscoveryInterface
+	// ttl is how long the openAPI schema should be considered valid
+	ttl time.Duration
+	// mu protects the gvkParser
+	mu sync.Mutex
+	// gvkParser retrieves the objectType for a given gvk
+	gvkParser *fieldmanager.GvkParser
+	// lastChecked is the last time we checked if the openAPI doc has changed.
+	lastChecked time.Time
+}
+
+// regenerateGVKParser builds the parser from the raw OpenAPI schema.
+func (c *gvkParserCache) regenerateGVKParser() error {
+	doc, err := c.discoveryClient.OpenAPISchema()
+	if err != nil {
+		return err
+	}
+	c.lastChecked = time.Now()
+	models, err := proto.NewOpenAPIData(doc)
+	if err != nil {
+		return err
+	}
+
+	gvkParser, err := fieldmanager.NewGVKParser(models, false)
+	if err != nil {
+		return err
+	}
+
+	c.gvkParser = gvkParser
+	return nil
+}
+
+// objectTypeForGVK retrieves the typed.ParseableType for a given gvk from the cache
+func (c *gvkParserCache) objectTypeForGVK(gvk schema.GroupVersionKind) (*typed.ParseableType, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.gvkParser != nil {
+		// if the ttl on the parser cache has expired,
+		// recheck the discovery client to see if the Open API schema has changed
+		if time.Now().After(c.lastChecked.Add(c.ttl)) {
+			c.lastChecked = time.Now()
```

Comment:
lastChecked is updated in two places which makes things confusing IMO

Reply:
fixed, now that gvkParser can't be nil